docs/glossary.rst

```diff
@@ -0,0 +1,274 @@
+
+    roughly map to terms like *Executor* and *Task* in some other frameworks.
+    Distinctions relate to the lifetime of the :term:`Context` instance, and the fact that
+    it owns both the work specification (including operation and data handles)
+    and the computing resources.
+    The :term:`Context` instance owns resources (on behalf of the client) that may
+    otherwise be owned directly by the client, and so its lifetime must span all
+    references to resources, operation handles, and data futures.
+
+.. glossary::
+
+    discovered task
+        A task that has become runnable, but was not already scheduled.
+        This term is intentionally vague as the requirements and constraints of
+        work management are explored. The primary usage refers to a task that
+        has been generated due to adaptations in the work flow. It may also
+        apply to tasks that may be scheduled opportunistically, or simply to
+        the change of state when a task's input dependencies have been met.
+
+.. _user classification:
+
+User classes and characteristics
+--------------------------------
+
+Users are assumed to be molecular science researchers using Python scripts to
+express and execute simulation and analysis work consisting of multiple
+simulation and analysis tasks, using software tools from multiple packages.
+
+Software tools are individually accessible as Python modules or as command line
+tools.
+
+Computational work may require multiple invocations (multiple HPC jobs) to complete.
+
+The following classes of user are not necessarily mutually exclusive.
+
+.. glossary::
+
+    basic user
+        A researcher writing a Python script to control standard software.
+
+    advanced user
+        A researcher who needs to integrate custom code into the scripted work.
+
+    pure Python user
+        All software necessary for the work is importable as Python modules.
+
+    mixed command line user
+        Some software is only accessible to Python by wrapping a command line driven tool.
+
+    compiled extension user
+        Some software necessary for the work requires compilation and/or installation
+        on the computing resource.
+
+    direct user
+    local user
+        Work is executed directly in the process(es) launched by the user.
+        Examples include a Python interpreter launched on the user's desktop or
+        a script launched with :command:`mpiexec` in a terminal window.
+
+    indirect user
+    remote user
+        A script run by the user dispatches serialized work through API-enabled
+        middleware for deserialization and execution outside of the user's
+        Python interpreter. In addition to remote execution systems, this class
+        may include adapters to *container* systems or job queuing systems,
+        whether or not the execution occurs on the same machine as the initial
+        Python interpreter.
+
+It is useful to define some classes of client code.
```

Comment:
@mrshirts writes:
> I think the term method needs to be defined before defining "iterative methods", "adaptive method", etc. And when saying "similar work", we should be using more precise language. "similar tasks", etc. or whatever is most correct. This would be a good section to start using precisely the language above.

Reply:
"method" might be replaceable with "use case". I can link "work" to the term and discussion elsewhere. I think further refinement is not possible without progress in other issues, but the current issue was requested in order to establish a base line for the other ongoing discussions.